### ISSUE TITLE
[feature request] large scale embedding for sparse features

### ISSUE THREAD

**Issue by candyzone** [tensorflow/tensorflow#19324]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux CentOS 7
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: 1.4.0 release
- **Python version**: python2.7
- **Bazel version (if compiling from source)**: 0.9.0
- **GCC/Compiler version (if compiling from source)**: gcc4.8.5
- **CUDA/cuDNN version**: -
- **GPU model and memory**: -
- **Exact command to reproduce**: -

In my scene, training data is very large, we have 10e10 unique values(string) for embedding in one sparse column. In short, model structure is “input->embedding->NNs”. I set hash_bucket_size to 15e10 for low hash collision. sample code:
```
col0=tf.contrib.layers.sparse_column_with_hash_bucket("feature_id", hash_bucket_size=1.5e11)
cols=[tf.feature_column.embedding_column(categorical_column=col0, dimension=8)]
```
in above example, Tensorflow new a Variable with shape [1.5e11, 8]
Some problems:
1. memory waste, more than 0.5e11\*8\*4Bytes memory not used. 
2. hash collision, though enlarging the `hash_bucket_size`, it might occurs conflict. I don't know how much it influence on model. 

Is there any suggestions in Tensorflow in this case ?
In my opinion: 
Define a new Variable for embedding, need not define the first demension. It will malloc memory for this Variable when the new value is embedding_lookup. It will solve the two problems mentioned above.

Thanks

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by candyzone** [comment]
@tensorflowbutler 
updated, thanks

**Comment by tatianashp** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

If you think we've misinterpreted a bug, please comment again with a clear explanation, as well as all of the information requested in the [issue template](https://github.com/tensorflow/tensorflow/issues/new). Thanks!

**Comment by caisq** [comment]
cc @ispirmustafa @roumposg, who may have more thoughts on this issue / feature request.

I question I'm wondering is whether hash-based embedding is appropriate for sparse features with number of unique values as large as 10e10.

**Comment by caisq** [comment]
(I am re-opening this issue because it might be a legitimate feature request.)

**Comment by ispirmustafa** [comment]
Hi @candyzone ,
I would use multiple hash keys to mitigate hash collisions. It will let model to differentiate collisions. 
It's implemented as `hash_keys` argument of `sparse_column_with_hash_bucket`.

**Comment by candyzone** [comment]
@caisq @ispirmustafa thanks, I see that API's usage, but it cannot solve my problem.
First is memory waste, we store embedding in Variable (a Tensor), only part of Variable is valid, the others wouldn't be lookup, this part of Variable is invalid, waste.
Second is Variable's first dimension(API's parameter hash_bucket_size) is fixed, so that it may cause hash collision and limit feature number in OnlineLearning case.
I think TensorFlow can provide a EmbeddingVariable (inherit from tf.Variable) that has an unknown first dimension and fixed other dimensions in Variable definition, it can implement using hash-base structure.

**Comment by cstur4** [comment]
I have the similar requirements and it is also an essential problem in online learning.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tatianashp** [comment]
We currently don't have anybody working on this. It would be great if you could help us by working on this and submitting a PR. Let us know if you need further clarification. Thanks!

**Comment by candyzone** [comment]
@yuefengz Here is my issue.

**Comment by yuefengz** [comment]
Reopen this issue since we get huge number of similar feature request in China. But we will still need to think about how to support it. Contribution is welcomed.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by llhe** [comment]
Anyone starts working on this? Looks like this is a common requirements for online learning.

**Comment by lovetimil** [comment]
> I have the similar requirements and it is also an essential problem in online learning.

me too

**Comment by chasex** [comment]
@candyzone TensorFlow use static computation graph, which means that TensorShape of inputs and outputs in each Op must be inferred at graph construction. So, I think it's not possible to dynamically grow embedding tensor's dim size in online learning case. (Correct me if I'm wrong..LOL)

**Comment by candyzone** [comment]
> @candyzone TensorFlow use static computation graph, which means that TensorShape of inputs and outputs in each Op must be inferred at graph construction. So, I think it's not possible to dynamically grow embedding tensor's dim size in online learning case. (Correct me if I'm wrong..LOL)

For dense layer, Variable's shape must be fixed, because the auto differential depend on the Variable‘s shape, gradient are dense Tensor.
But for sparse layer, gradients is list<pair<index, gradient>>, it allow us to make a SparseTensor that first dim is mutable, other dims is fixed. It seems that work well in Variable update.

**Comment by alextp** [comment]
In principle that's implementable, but the current tf Variable ops do not support this type of dynamic growth.

If you'd like to extend them to do so please write an RFC and we'll discuss there. I agree this is potentially very useful.

**Comment by tilakrayal** [comment]
@candyzone,
Related [PR](https://github.com/tensorflow/recommenders-addons/pull/16)  was closed. Could you please let us know if the issue still persists in the latest version. Thank you!

**Comment by chenli1994** [comment]
> Closing as stale. Please reopen if you'd like to work on this further.

.
